File: server/rails/bank-transfer.js
// ============================================================================
// Bank Wire / ACH / SEPA Payment Rail Connector
// Standard international wire, SEPA, and ACH transfer engine with ledger settlement
// ============================================================================
const crypto = require('crypto');
const db = require('../db');
const { recordIncome, recordSpending, getAccountBalance } = require('../ledger/ledger-core');

function ensureBankWireAccount(userId) {
  let account = db.get(
    "SELECT id, name, currency FROM accounts WHERE user_id = ? AND (name LIKE '%Wire%' OR id LIKE 'acc_wire_%')",
    [userId]
  );

  if (!account) {
    const accountId = 'acc_wire_' + crypto.randomBytes(4).toString('hex');
    db.run(
      `INSERT INTO accounts (id, user_id, name, type, account_class, currency, institution, initial_balance)
       VALUES (?, ?, 'Global Wire Account', 'checking', 'asset', 'USD', 'Lumen Clearing Bank', 0)`,
      [accountId, userId]
    );
    account = { id: accountId, name: 'Global Wire Account', currency: 'USD' };
  }

  return account;
}

/**
 * Initiates an outbound wire transfer (SWIFT, SEPA, ACH, or LOCAL)
 */
function initiateWireTransfer({ userId, iban, bic, recipientName, amountUsd, description = 'Wire Transfer', type = 'SWIFT' }) {
  const numericAmount = Number(amountUsd);
  if (isNaN(numericAmount) || numericAmount <= 0) {
    throw new Error('Wire amount must be a positive number in USD');
  }

  const bankAccount = ensureBankWireAccount(userId);
  const currentBalance = getAccountBalance(bankAccount.id);

  if (currentBalance < numericAmount) {
    throw new Error(`Insufficient funds for wire transfer. Available: $${currentBalance.toFixed(2)}, Requested: $${numericAmount.toFixed(2)}`);
  }

  const wireId = 'WIRE_' + type.toUpperCase() + '_' + Date.now() + '_' + crypto.randomBytes(3).toString('hex').toUpperCase();
  const railTxId = 'rail_' + crypto.randomUUID();

  // Atomically debit account in ledger
  const ledgerResult = recordSpending({
    userId,
    accountId: bankAccount.id,
    amount: numericAmount,
    category: 'Bank Wire Outward',
    payee: `${type} Wire to ${recipientName} (${iban.substring(0, 8)}...)`,
    referenceId: wireId,
    icon: 'landmark',
    tone: 'orange'
  });

  db.run(
    `INSERT INTO payment_rails_transactions (id, user_id, rail, direction, status, amount, currency, reference, phone_or_account, metadata_json, journal_entry_id)
     VALUES (?, ?, 'bank_transfer', 'outflow', 'pending', ?, 'USD', ?, ?, ?, ?)`,
    [
      railTxId,
      userId,
      numericAmount,
      wireId,
      iban,
      JSON.stringify({
        bic: bic || 'LUMXUS33',
        recipientName,
        transferType: type,
        description,
        bankAccountId: bankAccount.id
      }),
      ledgerResult.journalEntryId
    ]
  );

  return {
    success: true,
    wireId,
    railTxId,
    transferType: type,
    recipientName,
    iban,
    amountUsd: numericAmount,
    status: 'pending',
    message: `${type} wire transfer of $${numericAmount.toFixed(2)} to ${recipientName} submitted for clearing.`
  };
}

/**
 * Records an inbound wire transfer (Credit to user's clearing account)
 */
function receiveInboundWire({ userId, senderName, senderIban, amountUsd, reference, description = 'Inbound Wire Transfer' }) {
  const numericAmount = Number(amountUsd);
  if (isNaN(numericAmount) || numericAmount <= 0) {
    throw new Error('Inbound wire amount must be positive');
  }

  const bankAccount = ensureBankWireAccount(userId);
  const wireRef = reference || ('IN_WIRE_' + crypto.randomBytes(4).toString('hex').toUpperCase());
  const railTxId = 'rail_' + crypto.randomUUID();

  const ledgerResult = recordIncome({
    userId,
    accountId: bankAccount.id,
    amount: numericAmount,
    category: 'Bank Wire Inward',
    payee: `Inbound Wire from ${senderName}`,
    referenceId: wireRef,
    icon: 'landmark',
    tone: 'teal'
  });

  db.run(
    `INSERT INTO payment_rails_transactions (id, user_id, rail, direction, status, amount, currency, reference, phone_or_account, metadata_json, journal_entry_id)
     VALUES (?, ?, 'bank_transfer', 'inflow', 'completed', ?, 'USD', ?, ?, ?, ?)`,
    [
      railTxId,
      userId,
      numericAmount,
      wireRef,
      senderIban || 'N/A',
      JSON.stringify({ senderName, description, bankAccountId: bankAccount.id }),
      ledgerResult.journalEntryId
    ]
  );

  return {
    success: true,
    status: 'completed',
    reference: wireRef,
    senderName,
    amountUsd: numericAmount,
    newBalanceUsd: getAccountBalance(bankAccount.id),
    message: `Inbound wire of $${numericAmount.toFixed(2)} received from ${senderName} and credited to ledger.`
  };
}

/**
 * Handles clearing bank webhook update (e.g. wire transitions from 'pending' to 'completed' or 'failed')
 */
function handleBankWebhook(payload) {
  const wireId = payload?.wireId || payload?.reference;
  const newStatus = payload?.status?.toLowerCase();

  if (!wireId || !newStatus) {
    throw new Error('Invalid bank webhook payload: wireId and status are required');
  }

  const railTx = db.get(
    "SELECT * FROM payment_rails_transactions WHERE reference = ? AND rail = 'bank_transfer'",
    [wireId]
  );

  if (!railTx) {
    throw new Error(`Wire transaction not found for reference: ${wireId}`);
  }

  let metadata = {};
  try { metadata = JSON.parse(railTx.metadata_json); } catch (e) {}

  if (newStatus === 'completed' || newStatus === 'cleared') {
    db.run(
      "UPDATE payment_rails_transactions SET status = 'completed', updated_at = datetime('now') WHERE id = ?",
      [railTx.id]
    );

    return { success: true, wireId, status: 'completed', message: `Wire ${wireId} marked completed.` };
  } else if (newStatus === 'failed' || newStatus === 'rejected') {
    // If an outbound wire fails, credit back the user's account
    if (railTx.direction === 'outflow' && railTx.status === 'pending') {
      const bankAccount = ensureBankWireAccount(railTx.user_id);
      recordIncome({
        userId: railTx.user_id,
        accountId: bankAccount.id,
        amount: railTx.amount,
        category: 'Wire Reversal',
        payee: `Wire Refund (${wireId})`,
        referenceId: 'REV_' + wireId,
        icon: 'rotate-ccw',
        tone: 'teal'
      });
    }

    db.run(
      "UPDATE payment_rails_transactions SET status = 'failed', metadata_json = ?, updated_at = datetime('now') WHERE id = ?",
      [JSON.stringify({ ...metadata, failureReason: payload.reason || 'Rejected by intermediary bank' }), railTx.id]
    );

    return { success: true, wireId, status: 'failed', message: `Wire ${wireId} failed and funds reversed.` };
  }

  return { success: true, wireId, status: railTx.status };
}

function getWireStatus(wireId) {
  const tx = db.get("SELECT * FROM payment_rails_transactions WHERE reference = ? AND rail = 'bank_transfer'", [wireId]);
  if (!tx) return null;

  return {
    wireId: tx.reference,
    direction: tx.direction,
    status: tx.status,
    amountUsd: tx.amount,
    createdAt: tx.created_at,
    updatedAt: tx.updated_at,
    metadata: tx.metadata_json ? JSON.parse(tx.metadata_json) : {}
  };
}

module.exports = {
  ensureBankWireAccount,
  initiateWireTransfer,
  receiveInboundWire,
  handleBankWebhook,
  getWireStatus
};

File: src/rails/bank_transfer.rs
// ============================================================================
// Bank Wire Payment Rail Connector (SWIFT / SEPA / ACH) in Rust
// International & Local Bank Transfers, Wire Instructions, ISO 20022 messaging
// ============================================================================
use crate::ledger::core::{get_account_balance, record_income, record_spending, TransactionResult};
use rusqlite::{params, Connection};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WireTransferRequest {
    pub beneficiary_name: String,
    pub iban_or_account: String,
    pub bic_swift: String,
    pub bank_name: String,
    pub amount: Decimal,
    pub currency: String,
    pub wire_type: String, // "SWIFT", "SEPA", "ACH"
    pub reference: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WireTransferResult {
    pub success: bool,
    pub reference_number: String,
    pub wire_type: String,
    pub amount: Decimal,
    pub currency: String,
    pub estimated_settlement_hours: u32,
    pub debited_usd: Decimal,
    pub status: String,
    pub message: String,
}

pub fn ensure_bank_account(conn: &Connection, user_id: &str) -> String {
    let mut stmt = conn
        .prepare("SELECT id FROM accounts WHERE user_id = ? AND (type = 'checking' OR name LIKE '%Bank%')")
        .unwrap();

    let account_id: Option<String> = stmt.query_row([user_id], |row| row.get(0)).ok();

    if let Some(id) = account_id {
        id
    } else {
        let new_id = format!("acc_bank_{}", &Uuid::new_v4().to_string()[..8]);
        conn.execute(
            "INSERT INTO accounts (id, user_id, name, type, account_class, currency, institution, initial_balance)
             VALUES (?, ?, 'Primary Operating Bank Account', 'checking', 'asset', 'USD', 'JPMorgan Chase / Barclays', 0)",
            params![new_id, user_id],
        )
        .unwrap();
        new_id
    }
}

pub fn process_outward_wire(
    conn: &mut Connection,
    user_id: &str,
    req: WireTransferRequest,
) -> Result<WireTransferResult, String> {
    if req.amount <= Decimal::ZERO {
        return Err("Wire amount must be greater than zero".to_string());
    }

    let bank_account_id = ensure_bank_account(conn, user_id);
    let current_bal = get_account_balance(conn, &bank_account_id)?;

    if current_bal < req.amount {
        return Err(format!("Insufficient bank balance. Available: ${}, Requested: ${}", current_bal, req.amount));
    }

    let wire_ref = format!("WIRE_{}_{}", req.wire_type.to_uppercase(), &Uuid::new_v4().to_string()[..8].to_uppercase());
    let rail_tx_id = format!("rail_{}", Uuid::new_v4());

    let (settlement_hours, fee_usd) = match req.wire_type.to_uppercase().as_str() {
        "SWIFT" => (24, Decimal::new(25, 0)), // $25 SWIFT wire fee
        "SEPA" => (4, Decimal::new(2, 0)),    // $2 SEPA transfer
        "ACH" => (12, Decimal::ZERO),         // $0 ACH standard transfer
        _ => (24, Decimal::new(10, 0)),
    };

    let total_debit = req.amount + fee_usd;

    let ledger_res = record_spending(
        conn,
        user_id,
        &bank_account_id,
        total_debit,
        "Bank Wire Transfer Outward",
        &format!("{} wire to {} ({}) - BIC: {}", req.wire_type, req.beneficiary_name, req.iban_or_account, req.bic_swift),
        None,
        Some(wire_ref.clone()),
        "building-bank",
        "blue",
    )?;

    let meta = serde_json::json!({
        "beneficiary": req.beneficiary_name,
        "iban": req.iban_or_account,
        "swift": req.bic_swift,
        "bank": req.bank_name,
        "wireType": req.wire_type,
        "wireFeeUsd": fee_usd
    });

    conn.execute(
        "INSERT INTO payment_rails_transactions (id, user_id, rail, direction, status, amount, currency, reference, phone_or_account, metadata_json, journal_entry_id)
         VALUES (?, ?, 'bank_wire', 'outflow', 'completed', ?, ?, ?, ?, ?, ?)",
        params![
            rail_tx_id,
            user_id,
            total_debit.to_string().parse::<f64>().unwrap_or(0.0),
            req.currency,
            wire_ref,
            req.iban_or_account,
            meta.to_string(),
            ledger_res.journal_entry_id
        ],
    ).map_err(|e| e.to_string())?;

    Ok(WireTransferResult {
        success: true,
        reference_number: wire_ref,
        wire_type: req.wire_type.to_uppercase(),
        amount: req.amount,
        currency: req.currency.to_uppercase(),
        estimated_settlement_hours: settlement_hours,
        debited_usd: total_debit,
        status: "settled".to_string(),
        message: format!("{} transfer of {} {} dispatched to {} ({}) via ISO 20022 wire queue.", req.wire_type, req.amount, req.currency, req.beneficiary_name, req.bank_name),
    })
}
